src/wallet/orchard.h

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) 2021 The Zcash developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or https://www.opensource.org/licenses/mit-license.php .
+
+#ifndef ZCASH_WALLET_ORCHARD_H
+#define ZCASH_WALLET_ORCHARD_H
+
+#include "rust/orchard/keys.h"
+#include "rust/orchard/wallet.h"
+#include "zcash/address/orchard.hpp"
+
+class OrchardWallet
+{
+private:
+    std::unique_ptr<OrchardWalletPtr, decltype(&orchard_wallet_free)> inner;
+
+public:
+    OrchardWallet() : inner(orchard_wallet_new(), orchard_wallet_free) {}
+    OrchardWallet(OrchardWallet&& wallet_data) : inner(std::move(wallet_data.inner)) {}
+    OrchardWallet& operator=(OrchardWallet&& wallet)
+    {
+        if (this != &wallet) {
+            inner = std::move(wallet.inner);
+        }
+        return *this;
+    }
+
+    // OrchardWallet should never be copied
+    OrchardWallet(const OrchardWallet&) = delete;
+    OrchardWallet& operator=(const OrchardWallet&) = delete;
+
+    void CheckpointNoteCommitmentTree() {
+        orchard_wallet_checkpoint(inner.get());
+    }
+
+    bool RewindToLastCheckpoint() {
+        return orchard_wallet_rewind(inner.get());
+    }
+
+    /**
+     * Add notes that are decryptable with IVKs for which the wallet
+     * contains the full viewing key to the wallet.
+     */
+    bool AddNotesIfInvolvingMe(const CTransaction& tx) {
+        return orchard_wallet_add_notes_from_bundle(
+                inner.get(),
+                tx.GetHash().begin(),
+                tx.GetOrchardBundle().inner.get());
+    }
+
+    /**
+     * Append each Orchard note commitment from the specified block to the
+     * wallet's note commitment tree.
+     *
+     * Returns `false` if the caller attempts to insert a block out-of-order.
+     */
+    bool AppendNoteCommitments(const int nBlockHeight, const CBlock& block) {
+        uint32_t txidx = 0;
+        for (const CTransaction& tx : block.vtx) {
+            if (!orchard_wallet_append_bundle_commitments(
+                    inner.get(),
+                    (uint32_t) nBlockHeight,
+                    txidx,
+                    tx.GetHash().begin(),
+                    tx.GetOrchardBundle().inner.get()
+                    )) {
+                return false;
+            }
+            txidx += 1;
+        }
```

Comment:
The reason that I used `const int nBlockHeight` is that the rest of the C++ wallet code uses ints for block height; if we change this to uint32_t then we'll have to do the cast in wallet.cpp, and I thought that it was more idiomatic to make that an artifact of the FFI layer, which is what's getting wrapped here; since the Rust side uses u32, this is the correct place to perform the cast, and `uint32_t` is the correct type for the C header to use.

Reply:
Fine with me, but then the FFI should also use `uint32_t` instead of `unsigned int` in its header (i.e. it should be consistent with the Rust FFI method it is a header for).

Reply:
Thanks for the suggested change, I had missed the detail in the middle.